rust/kernel/src/file_operations.rs

```diff
@@ -41,16 +41,26 @@ pub enum SeekFrom {
     Current(i64),
 }
 
+fn unwrap_error<T>(r: KernelResult<T>) -> T
```

Comment:
Would it make more sense to do this as a macro, to avoid the need for closures for hte control flow?

Reply:
I am new to rust. What would the macro you have in mind do?

Reply:
It'd take the expression, and return the error integer, so you could use it like:

```rust
let f = err_to_kerror!(T::open())
```

in `open_callback`. It could probably use a better name :-)

Reply:
Does it mean that I'd have to wrap every call that can fail with err_to_kerror (or the name we choose)? If so, I'm not sure this would be an improvement over using the question mark; I guess it would for open_callback, but not for the others where there are more error paths.

I thought about having a macro that wrapped the whole function body so as to hide the closure, but I wasn't sure if the slight improvement in cleanness would outweigh the reduction in immediate understanding about what was going on. (Someone seeing the macro wouldn't know what's going on just by looking at the usage.)

Reply:
Yes, you'd wrap each call individually. To me this feels more idiomatic for Rust.

Reply:
I just found out that macro usage in rust does not have to be wrapped in parens, so I added a macro that hides the closure and makes the code look somewhat natural (without having to wrap each call individually). PTAL.

Reply:
For me it sounds like it should have been implemented as function of KernelResult.

Reply:
What should have been implemented as function of KernelResult?

Note that these functions are callbacks supplied to the rest of the kernel, we don't control the type.

Reply:
Oh, than nevermind. Just looked briefly on the functions

Reply:
I like the idea (with a better name for the macro), but what about forwarding the call to a safe, private function that returns `KernelResult` so that we can reduce the `unsafe` area? I am not sure how verbose it would be within the function bodies in these cases, though, considering what they do...

Reply:
Thanks for the review!

What name do you suggest for the macro? I'd be happy to change it :)

These functions can certainly be further improved, for example, read/write seem to assume that the whole buffer is read/written, which obviously not true. Additionally, they are meant to do what you suggest, that is, do some unsafe work marshalling arguments then call a safe rust-idiomatic function; note that they call T::open, T::READ, T::WRITE, etc.

What I'm trying to do with this patch is just replace the match-ok-err pattern with the use of question marks. I believe this removes clutter from the functions and makes it easier to read (and write) them -- this showcases rust's expressibility wrt error handling. I had the resist the temptation of making more changes.

Given the above, how do you feel about leaving this patch just doing that and then have additional changes on top of that in later patches?

[Some context: if you look at FileOperationsVtable::VTABLE below, you'll see that there are many additional file operations that can still be offered by drivers. For example, I need to implement unlocked_ioctl, compat_ioctl, poll, mmap, and flush for the driver I'm working on converting. The prospect writing these additional functions in the current style was not very inviting.]

Reply:
> What name do you suggest for the macro? I'd be happy to change it :)

One idea is using the same one as the function, `unwrap_error!`, since it is anyway what the macro in the end does. Another idea that conveys the fact that we are actually wrapping things into a block that returns a `KernelResult` is `with_Result!` (`with_KernelResult!` perhaps) or `from_Result` (`from_KernelResult!` perhaps).

> What I'm trying to do with this patch is just replace the match-ok-err pattern with the use of question marks

Yeah, what I was suggesting was writing a separate function instead so that we could not just use `?` but also have it as a safe function (and then using `unsafe` inside as needed). But I am not sure why the callbacks are marked `unsafe` to begin with -- is bindgen is marking the file operation pointers `unsafe`?

> I had the resist the temptation of making more changes.

Yes, we shouldn't get too clever. While this part (the `rust/` crates) is more or less OK to be fancy, for the drivers themselves we should aim to be as clear as possible.

> Given the above, how do you feel about leaving this patch just doing that and then have additional changes on top of that in later patches?

That's fine with me -- we are iterating here and the new code looks definitely cleaner. We can always improve things later on.

Reply:
> > What I'm trying to do with this patch is just replace the match-ok-err pattern with the use of question marks
> 
> Yeah, what I was suggesting was writing a separate function instead so that we could not just use `?` but also have it as a safe function (and then using `unsafe` inside as needed). But I am not sure why the callbacks are marked `unsafe` to begin with -- is bindgen is marking the file operation pointers `unsafe`?

That's right, bindgen is making the function pointers unsafe in file_operations, for example, open:
```
    pub open: ::core::option::Option<
        unsafe extern "C" fn(arg1: *mut inode, arg2: *mut file) -> c_types::c_int,
    >,
```

I considered doing what you're suggesting, though admittedly I wasn't thinking about safe/unsafe, I was just thinking about returning KernelResult. The problem was that it resulted in two copies of these ugly function declarations, with tiny definitions; I tried to avoid this by creating a `compose` function that would first call the KernelResult handler, then call unwrap_error to convert to the right return value. The problem with this was that each function has a different number of arguments, so a generic compose would require something like C++'s [parameter pack](https://en.cppreference.com/w/cpp/language/parameter_pack) or some trick that eludes me.

Then I considered the closure because it avoids having to repeat the function prototype/declaration by capturing the parameters. This solved the KernelResult issue I was trying to address, but as you point out still leaves us with a whole unsafe function -- which I think we should try to address in a different patch.

Perhaps we should consider, as you allude to, whether it's acceptable for bindgen to not tag these as unsafe.

Reply:
> whether it's acceptable for bindgen to not tag these as unsafe.

Yeah, we probably want to take a look at that at some point. I don't think there is a reason they need to be `unsafe` as long as they are defined by the Rust side.

If you want, create a one-line issue (we are using it as a TODO list more than anything).

Reply:
Good idea. Opened issue #43.